library/x509.c

```diff
@@ -565,117 +565,82 @@ int mbedtls_x509_get_name(unsigned char **p, const unsigned char *end,
     return ret;
 }
 
-static int x509_parse_int(unsigned char **p, size_t n, int *res)
-{
-    *res = 0;
-
-    for (; n > 0; --n) {
-        if ((**p < '0') || (**p > '9')) {
-            return MBEDTLS_ERR_X509_INVALID_DATE;
-        }
-
-        *res *= 10;
-        *res += (*(*p)++ - '0');
-    }
-
-    return 0;
-}
-
 static int x509_date_is_valid(const mbedtls_x509_time *t)
 {
-    int ret = MBEDTLS_ERR_X509_INVALID_DATE;
-    int month_len;
-
-    CHECK_RANGE(0, 9999, t->year);
-    CHECK_RANGE(0, 23,   t->hour);
-    CHECK_RANGE(0, 59,   t->min);
-    CHECK_RANGE(0, 59,   t->sec);
-
+    unsigned int month_days;
+    unsigned int year;
     switch (t->mon) {
         case 1: case 3: case 5: case 7: case 8: case 10: case 12:
-            month_len = 31;
+            month_days = 31;
             break;
         case 4: case 6: case 9: case 11:
-            month_len = 30;
+            month_days = 30;
             break;
         case 2:
-            if ((!(t->year % 4) && t->year % 100) ||
-                !(t->year % 400)) {
-                month_len = 29;
-            } else {
-                month_len = 28;
-            }
+            year = (unsigned int) t->year;
+            month_days = ((year & 3) || (!(year % 100)
+                                         && (year % 400)))
+                          ? 28 : 29;
             break;
         default:
-            return ret;
+            return MBEDTLS_ERR_X509_INVALID_DATE;
+    }
+
+    if ((unsigned int) (t->day - 1) >= month_days ||      /* (1 - days in month) */
+        /* (unsigned int)( t->mon - 1 ) >= 12 || */ /* (1 - 12) checked above */
```

Comment:
sorry to nit-pick, but shouldn't this be something like
```suggestion
        /* (unsigned int) (t->mon - 1) >= 12 || */  /* (1 - 12) checked above */
```